### DISCUSSION TITLE
How to create a new BinFont?

### DISCUSSION THREAD

**Discussion by wprinz** [arendst/Tasmota#22712] [Q&A] [answered]
I have migrated my Arduitouch device (ESP32 with an ILI9341 display) from my own program that I developed some time ago to Tasmota using Tasmota-display that I compiled with the #define USE_RAMFONT option to add additional fonts. I succeeded to use fonts that I downloaded from the BinFonts folder and then uploaded to the ESP32. 

Now I'd like to add other fonts. How can I e.g. convert gfx fonts that I got from here: http://oleddisplay.squix.ch to the binary format that is used by Tasmota? I tried to use fontconvert.py  without success. Any hint on how to create or convert new fonts that can be used as Tasmota BinFonts is appreciated.

**Comment by valdus2** [comment]
In that way...

https://github.com/arendst/Tasmota/discussions/21796#discussioncomment-10096164

**Comment by wprinz** [comment, marked as answer]
Many thanks for the immediate reply! I tested already [Script Editor](https://www.dropbox.com/sh/0us18ohui4c3k82/AACcVmpZ4AfpdrWE_MPFGmbma?dl=0) (For converting .h to .fnt files), but that failed with gfx.h files that I got using http://oleddisplay.squix.ch/ either with a "parameter error pop up" or by producing a "murks.fnt" file (murks is the german word for rubbish, good humor of the author :-)

Now I gave it a another try with gfx fonts that I converted using this site: https://rop.nl/truetype2gfx/.
The provided GFX fonts, as well as the GFX fonts that are converted from uploaded ttf files can be successfully converted with Script Editor to a .fnt file and they work well as a Tasmota BinFont / RAM font.

Thus the pipeline is:

- use  EITHER: https://rop.nl/truetype2gfx/ to create the gfx.h file from a ttf file
- OR use [FontConvert](https://github.com/KrisKasprzak/FontConvert) to convert a .ttf to a .h file by selecting specific character ranges, i.e. just numbers to reduce the file size
- convert the .h with Script Editor to a .fnt file (link above)
- upload to Tasmota-display that has been compiled with the #define USE_RAMFONT option
- activate your new font: displayText [zFfontFileName:f5c1l5]abcde  (use l5 = line 5 since RAM Fonts are often larger and thus the line position is not well calculated meaning that the appear to high on the display,

Many thanks to @valdus2  for pointing our the different alternatives and the initial pipeline.

**Reply by valdus2** [reply]
Maybe this is the recommended pipeline for you, but converting a font in your way (from the above links), uses all its characters,
and if you want to use a second or a third etc. font, there is no more ram or space to use (you can't unload a font).
By limiting your selection to conclude only the characters that you want, 
you can use much more typefaces at once, or use the ram and space for whatever you want.

**Reply by wprinz** [reply]
You are right thats the brute force method by converting the whole font. I've read somewhere that fonts are automaticaly unloaded when you load a new font with the Tasmota displayText [FmyNewFont.fnt:] cmd. Thus the limiting factor is the size of the file sytem of the ESP32. Anyhow using RAM economically is a good advice. I have added your recommendation to the pipeline above.
